File: 0128_longest_consecutive_sequence.py
from typing import List

# Solution #1

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        mapper = {}
        max_length = 0

        for num in nums:
            if num in mapper.keys():
                continue
            mapper[num] = 0
            num_left = mapper.get(num-1, 0)
            num_right = mapper.get(num+1, 0)
            count = num_left+num_right+1
            mapper[num-num_left] = count
            mapper[num+num_right] = count
            print(mapper)
            max_length = max(max_length, count)
        
        return max_length
    
def run_test(nums):
    s = Solution()
    val = s.longestConsecutive(nums)
    return val

if __name__ == "__main__":
    nums = [0,3,7,2,5,8,4,6,0,1]
    val = run_test(nums)
    print(val)

File: cpp/0128_longest_consecutive_sequence.cpp
#include <cassert>
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  int longestConsecutive(vector<int>& nums) {
    int max_length{};

    unordered_map<int, int> hash_map{};

    for (auto& num : nums) {
      if (hash_map.find(num) != hash_map.end()) {
        continue;
      }
      hash_map[num] = 0;
      auto num_left =
          (hash_map.find(num - 1) != hash_map.end()) ? hash_map[num - 1] : 0;
      auto num_right =
          (hash_map.find(num + 1) != hash_map.end()) ? hash_map[num + 1] : 0;
      int count = num_right + num_left + 1;
      hash_map[num - num_left] = count;
      hash_map[num + num_right] = count;
      std::cout << num << " ";
      for (auto& i : hash_map) {
        std::cout << "{" << i.first << "," << i.second << "}, ";
      }
      std ::cout << std::endl;
      max_length = max(max_length, count);
    }

    return max_length;
  }
};

int main() {
  Solution s;
  std::vector<int> nums = {4,  0,  -4, -2, 2, 5, 2,  0, -8, -8, -8,
                           -8, -1, 7,  4,  5, 5, -4, 6, 6,  -3};

  auto out = s.longestConsecutive(nums);
  std::cout << out;
  // assert(out == 4);
}
